src/command-parser/commandParser.ts

```diff
@@ -17,36 +16,29 @@ export enum CommandType {
   CATALYST_VOTING_KEY_REGISTRATION_METADATA = 'catalyst.voting-key-registration-metadata',
 }
 
-const makeParser = () => {
-  const initParser = (parser: any, config: any) => {
-    const isCommand = (str: string) => !str.startsWith('--')
-    const commandType = (parent: string, current: string) => (parent ? `${parent}.${current}` : current)
-    parser.set_defaults({ parser })
-    const subparsers = parser.add_subparsers()
-
-    Object.keys(config).reduce((acc, key) => {
-      if (isCommand(key)) {
-        const subparser = acc.add_parser(key)
-        subparser.set_defaults({ command: commandType(parser.get_default('command'), key) })
-        initParser(subparser, config[key])
-      } else {
-        parser.add_argument(key, config[key])
-      }
-      return acc
-    }, subparsers)
-
-    return parser
-  }
-
-  return initParser(new ArgumentParser(
-    {
-      description: 'Command line tool for ledger/trezor transaction signing',
-      prog: 'cardano-hw-cli',
-    },
-  ), parserConfig)
+const initParser = (parser: ArgumentParser | ArgumentGroup, config: any): void => {
+  const MUTUALLY_EXCLUSIVE_GROUP_KEY = '_mutually-exclusive-group'
+  const isMutuallyExclusiveGroup = (str: string) => str.startsWith(MUTUALLY_EXCLUSIVE_GROUP_KEY)
+  const isOneOfGroupRequired = (str: string) => str.endsWith(`${MUTUALLY_EXCLUSIVE_GROUP_KEY}-required`)
```

Comment:
Shouldn't it be at the start of the string? But we are using `endsWith`.

Reply:
Don't we have some unit tests for command parser? This definitely deserves a test, one for optional, one for required mutually exclusive group (maybe having two groups, too).

Reply:
> Shouldn't it be at the start of the string? But we are using `endsWith`.

Thanks for noticing! That was really a stupid mistake. (I remember testing it manually, but perhaps I forgot to rerun `yarn build-js` or something.)

> Don't we have some unit tests for command parser? This definitely deserves a test, one for optional, one for required mutually exclusive group (maybe having two groups, too).

[We do](https://github.com/vacuumlabs/cardano-hw-cli/blob/develop/test/unit/commandParser/commandParser.js), but they seem more like integration tests for command parsing - the tests are verifying whether the entire input from command line matches expected `ParsedArguments` after parsing, not whether individual parsers work as expected. I don't think developing an entire new command parser testing approach is worth it (considering that the CLI API and implementation doesn't change that much).

However, I noticed that `address show` test is missing, so I added at least that. I tried to add a test for both payment path and script hash present (which violates the mutual exclusiveness), but it turns out that when `argparse` founds a problem it terminates the process instead of throwing an error. There is an option `exit_on_error` which can be set to `false`, but I couldn't get it working with mutually exclusive groups correctly. Again, I'm not sure it's worth the effort and added complexity in code.

Reply:
I don't think we need to test individual parsers.

At this point, I would leave this as an open github issue. We can return to it someday, perhaps after argparse will start working as expected :) Just write down a concise summary of why you couldn't make it work in the issue.

Reply:
I agree, issue added https://github.com/vacuumlabs/cardano-hw-cli/issues/111